src/consumer/mod.rs

```diff
@@ -254,10 +250,10 @@ impl Consumer {
                     // certain errors and retry the fetch operation
                     // transparently for the caller.
                     let data = match p.data() {
-                        // XXX need to prevent updating fetch_offsets in case we're gonna fail here
-                        Err(ref e) => return Err(e.clone()),
-                        Ok(ref data) => data,
+                        Ok(d) => d ,
+                        Err( e) => {  bail!("kafka failed") } //bail!(e) } // zlb: not ideal, would much rather be able to throw the actual error, but need to wire up copy and dependent errors don't do that
```

Comment:
@estk This is where I had the biggest challenge. I couldn't actually return/throw the error since it seems like it wanted to take ownership of the error, but causing the loop to take ownership of `resps` but `resps` is used below to form the new object. 
There are a couple ways to change or fix this imo:
- Make error copyable (which would be challenging and force us to do some restructuring), this is hard because there are a couple of nested error types that do not support the copy trait. What is actually probably the right answer, is to have a KafkaProtocolError or something akin to that. Then this can be a restructured to return a `Result<data, KafkaProtocolError>` which can be copied.
- The other alternative is to mutate the data as it goes, instead of only having one copy of the data in place. Although I do overall like this methodology, but it makes the ownership challenging. 

Hrmm as I typed this up I think my first option is a pretty reasonable solution, just didn't think about it enough (or had to sleep on it) earlier on. ;)